File: inference-v3/src/engine/models/qwen35/preparation.py
"""Qwen's CPU image preparation and exact decoder-span interpretation."""

import json
from dataclasses import dataclass
from hashlib import sha256
from importlib.metadata import version
from pathlib import Path

import numpy as np
from pydantic import PositiveInt

from engine.data import Record, TokenId
from engine.inputs.layout import ConditioningIdentity, InputLayout, InputPosition, InputSpan
from engine.inputs.media import PreparedMedia, PreparedTensor
from engine.models.qwen35.inputs import InputPlan


class ImageGeometry(Record):
    channels: PositiveInt
    temporal_patch: PositiveInt
    patch: PositiveInt
    merge: PositiveInt
    image_token: TokenId
    start_token: TokenId
    end_token: TokenId

    @property
    def patch_width(self):
        return self.channels * self.temporal_patch * self.patch**2


@dataclass(frozen=True)
class ImagePatches:
    identity: ConditioningIdentity
    grid: tuple[int, int, int]
    pixels: PreparedTensor


@dataclass(frozen=True)
class PreparedInput:
    plan: InputPlan
    images: tuple[ImagePatches, ...]


def spatial_controls(grid: tuple[int, int, int], merge: int, table_side: int):
    """Geometry-only rotary positions and bilinear lookup controls in patch order."""
    t, h, w = grid
    if t != 1 or merge <= 0 or table_side <= 0 or min(h, w) < merge or h % merge or w % merge:
        raise ValueError("invalid image geometry for spatial controls")
    row, col = np.meshgrid(np.arange(h), np.arange(w), indexing="ij")

    def patch_order(value):
        return value.reshape(h // merge, merge, w // merge, merge).transpose(0, 2, 1, 3).reshape(-1)

    coordinates = np.stack((patch_order(row), patch_order(col)), axis=-1).astype(np.int32)
    # align_corners=True, as in the bound Qwen processor/model contract.
    y = np.linspace(0, table_side - 1, h, dtype=np.float32)
    x = np.linspace(0, table_side - 1, w, dtype=np.float32)
    y0, x0 = y.astype(np.int32), x.astype(np.int32)
    y1, x1 = np.minimum(y0 + 1, table_side - 1), np.minimum(x0 + 1, table_side - 1)
    dy, dx = y - y0, x - x0
    indices, coefficients = [], []
    for yi, yw in ((y0, 1 - dy), (y1, dy)):
        for xi, xw in ((x0, 1 - dx), (x1, dx)):
            indices.append(patch_order(yi[:, None] * table_side + xi[None, :]).astype(np.int32))
            coefficients.append(patch_order(yw[:, None] * xw[None, :]).astype(np.float32)[:, None])
    return coordinates, tuple(indices), tuple(coefficients)


def interpret(
    tokens: tuple[TokenId, ...], media: PreparedMedia, *, processor: str, geometry: ImageGeometry
) -> PreparedInput:
    """Validate all media and placeholders before constructing executable input.

    Physical token offsets and rotary coordinates deliberately advance by
    different amounts across an image. Content identity includes preparation.
    """
    if media.processor != processor:
        raise ValueError("image processor differs from the bound artifact")
    fields = {tensor.name: tensor for tensor in media.tensors}
    if set(fields) != {"pixel_values", "image_grid_thw"}:
        raise ValueError("Qwen prepared tensors differ from the image contract")
    pixels, grids = fields["pixel_values"].array(), fields["image_grid_thw"].array()
    if (
        pixels.dtype != np.dtype("<f4")
        or pixels.ndim != 2
        or pixels.shape[1] != geometry.patch_width
        or grids.dtype != np.dtype("<i8")
        or grids.ndim != 2
        or grids.shape[1] != 3
        or not 1 <= grids.shape[0] <= 16
        or not np.isfinite(pixels).all()
    ):
        raise ValueError("invalid Qwen image tensor geometry or values")
    merge = geometry.merge
    # Python integer products cannot wrap before checking the payload extent.
    sizes = grids.tolist()
    if any(t != 1 or h < merge or w < merge or h % merge or w % merge for t, h, w in sizes):
        raise ValueError("image grids require one frame and merge-aligned spatial dimensions")
    if sum(t * h * w for t, h, w in sizes) != len(pixels):
        raise ValueError("image grids must cover their pixel patches exactly")
    coordinates = []
    images, spans = [], []
    cursor = pixel_start = rotary = 0
    for t, h, w in sizes:
        patches, count = t * h * w, t * h * w // merge**2
        try:
            start = tokens.index(geometry.image_token, cursor)
        except ValueError as error:
            raise ValueError("image has no corresponding prompt span") from error
        end = start + count
        if (
            start == 0
            or tokens[start - 1] != geometry.start_token
            or tokens[start:end] != (geometry.image_token,) * count
            or end >= len(tokens)
            or tokens[end] != geometry.end_token
        ):
            raise ValueError("image prompt span differs from feature geometry")
        coordinates.extend((p, p, p) for p in range(rotary, rotary + start - cursor))
        base = rotary + start - cursor
        coordinates.extend(
            (base + frame, base + row, base + col)
            for frame in range(t)
            for row in range(h // merge)
            for col in range(w // merge)
        )
        values = PreparedTensor.from_array(
            "pixel_values", pixels[pixel_start : pixel_start + patches]
        )
        digest = sha256(processor.encode())
        digest.update(np.asarray((t, h, w), dtype="<i8").tobytes())
        digest.update(values.data)
        identity = ConditioningIdentity(digest.hexdigest())
        images.append(ImagePatches(identity, (t, h, w), values))
        spans.append(
            InputSpan(start=InputPosition(start), end=InputPosition(end), identity=identity)
        )
        rotary = base + max(t, h // merge, w // merge)
        cursor, pixel_start = end, pixel_start + patches
    if geometry.image_token in tokens[cursor:]:
        raise ValueError("image prompt span has no source pixels")
    coordinates.extend((p, p, p) for p in range(rotary, rotary + len(tokens) - cursor))
    return PreparedInput(
        InputPlan(
            tokens=tokens,
            layout=InputLayout(count=len(tokens), spans=tuple(spans)),
            coordinates=tuple(coordinates),
            continuation=rotary + len(tokens) - cursor,
        ),
        tuple(images),
    )


def preparation_identity(directory: Path) -> str:
    digest = sha256(b"qwen35-image-preparation-v1")
    for content in (
        (directory / "config.json").read_bytes(),
        (directory / "preprocessor_config.json").read_bytes(),
        version("transformers").encode(),
        version("pillow").encode(),
    ):
        digest.update(len(content).to_bytes(8, "little"))
        digest.update(content)
    return digest.hexdigest()


class ImagePreparation:
    """Artifact-bound PIL processor; request data never becomes processor state."""

    def __init__(self, directory: Path, pieces: tuple[str, ...]):
        from transformers.models.auto.image_processing_auto import AutoImageProcessor

        directory = directory.expanduser().resolve()
        config_bytes = (directory / "config.json").read_bytes()
        config = json.loads(config_bytes)
        vision = config["vision_config"]
        self.geometry = ImageGeometry(
            channels=vision["in_channels"],
            temporal_patch=vision["temporal_patch_size"],
            patch=vision["patch_size"],
            merge=vision["spatial_merge_size"],
            image_token=config["image_token_id"],
            start_token=config["vision_start_token_id"],
            end_token=config["vision_end_token_id"],
        )
        self.identity = preparation_identity(directory)
        self.image_token = pieces[self.geometry.image_token]
        self.marker = "".join(
            pieces[token]
            for token in (
                self.geometry.start_token,
                self.geometry.image_token,
                self.geometry.end_token,
            )
        )
        self.processor = AutoImageProcessor.from_pretrained(
            directory, local_files_only=True, trust_remote_code=False, backend="pil"
        )

    def prepare(self, text: str, images: tuple) -> tuple[str, PreparedMedia]:
        parts = text.split(self.image_token)
        if not images or len(parts) != len(images) + 1:
            raise ValueError("prompt must contain one placeholder per source image")
        data = self.processor(images=list(images), return_tensors="np")
        if set(data) != {"pixel_values", "image_grid_thw"}:
            raise ValueError("image processor output differs from the bound contract")
        grids = np.asarray(data["image_grid_thw"])
        if grids.shape != (len(images), 3):
            raise ValueError("image processor grids differ from source count")
        counts = [int(t) * int(h) * int(w) // self.geometry.merge**2 for t, h, w in grids]
        media = PreparedMedia(
            self.identity,
            tuple(
                PreparedTensor.from_array(name, np.asarray(value)) for name, value in data.items()
            ),
        )
        expanded = parts[0] + "".join(
            self.image_token * count + part for count, part in zip(counts, parts[1:], strict=True)
        )
        return expanded, media

File: inference-v4/engine/src/models/qwen35/preparation.rs
//! Qwen image tensor interpretation. This validates processor output and prompt
//! spans without creating device resources or running an image encoder.
use crate::inputs::{
    media::{DType, PreparedMedia, PreparedTensor},
    BoundaryRule, InputLayout, InputSpan, TokenId,
};
use sha2::{Digest, Sha256};

#[derive(Clone, Debug)]
pub struct ImageGeometry {
    pub channels: usize,
    pub temporal_patch: usize,
    pub patch: usize,
    pub merge: usize,
    pub image_token: TokenId,
    pub start_token: TokenId,
    pub end_token: TokenId,
}
impl ImageGeometry {
    pub fn patch_width(&self) -> Result<usize, String> {
        if [self.channels, self.temporal_patch, self.patch, self.merge].contains(&0)
            || [self.image_token, self.start_token, self.end_token]
                .iter()
                .any(|id| id.0 > i32::MAX as u32)
            || self.image_token == self.start_token
            || self.image_token == self.end_token
            || self.start_token == self.end_token
        {
            return Err("invalid Qwen image geometry".into());
        }
        self.channels
            .checked_mul(self.temporal_patch)
            .and_then(|n| n.checked_mul(self.patch))
            .and_then(|n| n.checked_mul(self.patch))
            .ok_or("image patch width overflows".into())
    }
}
#[derive(Clone, Debug)]
pub struct InputPlan {
    tokens: Vec<TokenId>,
    layout: InputLayout,
    coordinates: Vec<[i32; 3]>,
    continuation: usize,
}
impl InputPlan {
    pub fn text(tokens: Vec<TokenId>) -> Result<Self, String> {
        if tokens.iter().any(|id| id.0 > i32::MAX as u32) {
            return Err("input token exceeds int32 domain".into());
        }
        let layout = InputLayout::new(tokens.len(), vec![])?;
        Ok(Self {
            continuation: tokens.len(),
            tokens,
            layout,
            coordinates: vec![],
        })
    }
    pub fn tokens(&self) -> &[TokenId] {
        &self.tokens
    }
    pub fn layout(&self) -> &InputLayout {
        &self.layout
    }
    pub fn continuation(&self) -> usize {
        self.continuation
    }
    pub fn rotary(&self, position: usize, count: usize) -> Result<Vec<[i32; 3]>, String> {
        let end = position
            .checked_add(count)
            .filter(|&n| n <= i32::MAX as usize)
            .ok_or("rotary range exceeds int32 domain")?;
        (position..end)
            .map(|index| {
                if index < self.tokens.len() {
                    Ok(self
                        .coordinates
                        .get(index)
                        .copied()
                        .unwrap_or([index as i32; 3]))
                } else {
                    let value = self
                        .continuation
                        .checked_add(index - self.tokens.len())
                        .and_then(|n| i32::try_from(n).ok())
                        .ok_or("rotary continuation exceeds int32 domain")?;
                    Ok([value; 3])
                }
            })
            .collect()
    }
}
#[derive(Clone, Debug)]
pub struct ImagePatches {
    pub identity: String,
    pub grid: [usize; 3],
    pub pixels: PreparedTensor,
}
#[derive(Clone, Debug)]
pub struct PreparedInput {
    pub plan: InputPlan,
    pub images: Vec<ImagePatches>,
}
pub fn interpret(
    tokens: Vec<TokenId>,
    media: &PreparedMedia,
    processor: &str,
    geometry: &ImageGeometry,
) -> Result<PreparedInput, String> {
    let width = geometry.patch_width()?;
    let mut plan = InputPlan::text(tokens)?;
    if media.processor() != processor {
        return Err("image processor differs from the bound artifact".into());
    }
    if media.tensors().len() != 2 {
        return Err("Qwen prepared tensors differ from the image contract".into());
    }
    let pixels = media
        .tensors()
        .iter()
        .find(|t| t.name() == "pixel_values")
        .ok_or("missing image pixels")?;
    let grids = media
        .tensors()
        .iter()
        .find(|t| t.name() == "image_grid_thw")
        .ok_or("missing image grids")?;
    if pixels.dtype() != DType::F32
        || pixels.shape().len() != 2
        || pixels.shape()[1] != width
        || grids.dtype() != DType::I64
        || grids.shape().len() != 2
        || grids.shape()[1] != 3
        || grids.shape()[0] > 16
        || pixels
            .data()
            .chunks_exact(4)
            .any(|b| !f32::from_le_bytes(b.try_into().unwrap()).is_finite())
    {
        return Err("invalid Qwen image tensor geometry or values".into());
    }
    let merge = geometry.merge;
    let merge_area = merge
        .checked_mul(merge)
        .ok_or("image merge area overflows")?;
    let sizes = grids
        .data()
        .chunks_exact(24)
        .map(|row| {
            let mut grid = [0usize; 3];
            for (index, bytes) in row.chunks_exact(8).enumerate() {
                grid[index] = usize::try_from(i64::from_le_bytes(bytes.try_into().unwrap()))
                    .map_err(|_| "negative image grid")?;
            }
            let [t, h, w] = grid;
            if t != 1 || h < merge || w < merge || h % merge != 0 || w % merge != 0 {
                return Err("image grids require one frame and merge-aligned dimensions");
            }
            Ok(grid)
        })
        .collect::<Result<Vec<_>, &str>>()?;
    let total = sizes
        .iter()
        .try_fold(0usize, |n, &[_, h, w]| {
            h.checked_mul(w).and_then(|rows| n.checked_add(rows))
        })
        .ok_or("image grid extent overflows")?;
    if total != pixels.shape()[0] {
        return Err("image grids must cover pixel patches exactly".into());
    }
    let mut images = Vec::new();
    let mut spans = Vec::new();
    let (mut cursor, mut pixel_start, mut rotary) = (0usize, 0usize, 0usize);
    for grid in sizes {
        let [t, h, w] = grid;
        let patches = h * w; // checked by the aggregate extent above
        let count = patches / merge_area;
        let start = plan.tokens[cursor..]
            .iter()
            .position(|&token| token == geometry.image_token)
            .map(|i| cursor + i)
            .ok_or("image has no corresponding prompt span")?;
        let end = start
            .checked_add(count)
            .filter(|&end| end < plan.tokens.len())
            .ok_or("image prompt span exceeds input")?;
        if start == 0
            || plan.tokens[start - 1] != geometry.start_token
            || plan.tokens[end] != geometry.end_token
            || plan.tokens[start..end]
                .iter()
                .any(|&token| token != geometry.image_token)
        {
            return Err("image prompt span differs from feature geometry".into());
        }
        let base = rotary
            .checked_add(start - cursor)
            .ok_or("rotary extent overflows")?;
        let continuation = base
            .checked_add(t.max(h / merge).max(w / merge))
            .filter(|&n| n <= i32::MAX as usize)
            .ok_or("image rotary extent exceeds int32 domain")?;
        plan.coordinates
            .extend((rotary..base).map(|p| [p as i32; 3]));
        for row in 0..h / merge {
            for col in 0..w / merge {
                plan.coordinates
                    .push([base as i32, (base + row) as i32, (base + col) as i32]);
            }
        }
        let byte_start = pixel_start * width * 4;
        let byte_end = (pixel_start + patches) * width * 4;
        let values = PreparedTensor::new(
            "pixel_values".into(),
            DType::F32,
            vec![patches, width],
            pixels.data()[byte_start..byte_end].to_vec(),
        )?;
        let mut digest = Sha256::new();
        digest.update(processor.as_bytes());
        for &value in &grid {
            digest.update((value as i64).to_le_bytes());
        }
        digest.update(values.data());
        let identity: String = digest
            .finalize()
            .iter()
            .map(|b| format!("{b:02x}"))
            .collect();
        spans.push(InputSpan {
            start,
            end,
            identity: identity.clone(),
            boundaries: BoundaryRule::Causal,
            language_history: false,
        });
        images.push(ImagePatches {
            identity,
            grid,
            pixels: values,
        });
        rotary = continuation;
        cursor = end;
        pixel_start += patches;
    }
    if plan.tokens[cursor..].contains(&geometry.image_token) {
        return Err("image prompt span has no source pixels".into());
    }
    plan.continuation = rotary
        .checked_add(plan.tokens.len() - cursor)
        .filter(|&n| n <= i32::MAX as usize)
        .ok_or("rotary continuation exceeds int32 domain")?;
    plan.coordinates
        .extend((rotary..plan.continuation).map(|p| [p as i32; 3]));
    plan.layout = InputLayout::new(plan.tokens.len(), spans)?;
    Ok(PreparedInput { plan, images })
}

/// Geometry-only encoder controls in merge-group patch order. Position-table
/// interpolation uses align_corners, matching the Qwen processor/model contract.
pub struct SpatialControls {
    pub coordinates: Vec<[i32; 2]>,
    pub indices: [Vec<i32>; 4],
    pub coefficients: [Vec<f32>; 4],
}
pub fn spatial_controls(
    grid: [usize; 3],
    merge: usize,
    table_side: usize,
) -> Result<SpatialControls, String> {
    let [t, h, w] = grid;
    if t != 1
        || merge == 0
        || table_side == 0
        || h < merge
        || w < merge
        || h % merge != 0
        || w % merge != 0
    {
        return Err("invalid image geometry for spatial controls".into());
    }
    table_side
        .checked_mul(table_side)
        .filter(|&n| n <= i32::MAX as usize)
        .ok_or("position table exceeds int32 addressability")?;
    let rows = h
        .checked_mul(w)
        .filter(|&rows| rows <= crate::inputs::media::MAX_PREPARED_BYTES / 40)
        .ok_or("spatial controls exceed prepared byte capacity")?;
    let mut result = SpatialControls {
        coordinates: Vec::with_capacity(rows),
        indices: std::array::from_fn(|_| Vec::with_capacity(rows)),
        coefficients: std::array::from_fn(|_| Vec::with_capacity(rows)),
    };
    let axis = |position: usize, extent: usize| {
        // np.linspace(..., dtype=float32) rounds before interpolation weights
        // are formed. Its endpoint is exact, including the singleton case.
        let value = if extent == 1 {
            0.0
        } else {
            (position as f64 * ((table_side - 1) as f64 / (extent - 1) as f64)) as f32
        };
        let low = value as usize;
        (
            low,
            (low + 1).min(table_side - 1),
            value as f64 - low as f64,
        )
    };
    for block_row in 0..h / merge {
        for block_col in 0..w / merge {
            for inner_row in 0..merge {
                for inner_col in 0..merge {
                    let row = block_row * merge + inner_row;
                    let col = block_col * merge + inner_col;
                    result.coordinates.push([row as i32, col as i32]);
                    let (y0, y1, dy) = axis(row, h);
                    let (x0, x1, dx) = axis(col, w);
                    for (index, (y, x, coefficient)) in [
                        (y0, x0, (1.0 - dy) * (1.0 - dx)),
                        (y0, x1, (1.0 - dy) * dx),
                        (y1, x0, dy * (1.0 - dx)),
                        (y1, x1, dy * dx),
                    ]
                    .into_iter()
                    .enumerate()
                    {
                        result.indices[index].push((y * table_side + x) as i32);
                        result.coefficients[index].push(coefficient as f32);
                    }
                }
            }
        }
    }
    Ok(result)
}
